### ISSUE TITLE
Voodoo 1/2/3 highly unstable (crashes 86Box or glitches and artifacts) on Apple Silicon

### ISSUE THREAD

**Issue by acordier16** [86Box/86Box#2764]
**Describe the bug**
Voodoo 1/2/3 cards seem to all have issues when using Apple Silicon architectures (I'm using Macbook Pro M2).

I've tried various combinations of chipset / CPU / Voodoo as well as most available drivers for each and:
- either the card works (with good performance) but has quite annoying artifacts and glitches
- either 86Box crashes a couple seconds/minutes after I run some 3D-accelerated software (3D Max 99, Star Wars Pod Racer, etc.)

I don't seem to be the only one given the comment below [this youtube video](https://www.youtube.com/watch?v=W50CtG0Lvls).

**To Reproduce**
1. Install Windows 98 via 86Box on Silicon Macbook, using Socket 7 / Slot 1 / Slot 2 chipset and CPU
2. Select Voodoo Graphics / Voodoo 2 (on top of S3 Virge DX) or Voodoo 3 2000
3. Install drivers and DirectX 7.0a
4. Run 3D Max 99 demo or benchmark (or any other Direct3D accelerated software)

**Expected behavior**
- Either glitches straight away
- or an 86Box crash after a couple seconds / minutes

**Screenshots**
Will try to take some later or shoot a video.

**Desktop (please complete the following information):**
 - OS: MacOS Monterey 12.5.1
 - 86Box version: 86Box v3.7.1 [build 4032] [730af4dd5]
 - Build information: latest MacOS build

**Additional context**
Add any other context about the problem here. If you are using an Optimized build, make sure to try the regular build too before filing a bug report!

**Comment by Cacodemon345** [comment]
We do not have a recompiler for Voodoo on ARM64 hosts unlike x86, so crashes and major glitches are expected to be common.

**Comment by acordier16** [comment]
> We do not have a recompiler for Voodoo on ARM64 hosts unlike x86, so crashes and major glitches are expected to be common.

Thanks for the precision, clarifies the problem. Do you know if any advance is planned on the topic?

**Comment by Cacodemon345** [comment]
We don't have anyone who can write a Voodoo recompiler for ARM64 hosts. If we do eventually have someone to do so, we may be able to get one, and therefore reduce the glitches and bugs.

**Comment by acordier16** [comment]
I've delved a bit more on the topic and compiled 86Box myself. It looks like that the fatal error is related to the Voodoo FIFO queue code, as mentioned somewhere in the Discord. Specifically, this line: https://github.com/86Box/86Box/blob/master/src/video/vid_voodoo_fifo.c#L276

Interestingly, I may have found a simple workaround. Changing the _fatal_ into a simple _printf_ to catch the error seems to have solved the crash issue for me. At least I manage to get a Voodoo Graphics running on my ARM64 Mac without any issue (no crash so far, no artifacts), which is not the case if I compile it without the fix. 🥳 The fact that it works may also be due to the particular config I chose, so I'll be sure to try out with other Voodoo cards and chipsets. What is weird though is that I don't seem to get anything printed anyway (maybe the printf is not redirected towards the std output).

If I understand correctly, a JIT recompiler targeted at ARM64 architectures for the Voodoo interpreter would simply help running the hot parts of the code better by performing dynamic compilation and avoid running each instruction to the interpreter one by one. If the Voodoo interpreter was bug-free, it would maybe run slow or be somewhat non-synced with the CPU (which by the way is not the case from what I experience on the M2 Macbook Pro), but not crash. So maybe there's an issue in the instructions handling of the interpreter? Does it also crashes on x86 if you disable the Voodoo dynamic recompilation?  That may be a way to experiment more.

I'd be happy to hear your thoughts!

NB: I compiled with the following flags:
- NEW_DYNAREC=ON
- CMAKE_TOOLCHAIN_FILE=./cmake/flags-gcc-aarch64.cmake
- CMAKE_BUILD_TYPE=Debug
